verilog/analysis/checkers/undersized_binary_literal_rule.cc

```diff
@@ -55,65 +58,108 @@ const char UndersizedBinaryLiteralRule::kTopic[] = "number-literals";
-            CHECK(number.ok)
-                << "Expecting valid numeric literal from lexer, but got: "
-                << digits_text;
-            // Detect binary values, whose literal width is shorter than the
-            // declared width.
-            // Allow 'b0 and 'b? as an exception.
-            CHECK_EQ(number.base,
-                     'b');  // guaranteed by matching TK_BinBase
-            if (width > number.literal.length() && number.literal != "0" &&
-                number.literal != "?") {
-              violations_.insert(LintViolation(
-                  digits_leaf->get(),
-                  FormatReason(width_text, base_text, digits_text), context));
-            }
-          }  // else width is not constant, so ignore
-        }
-      }
-    }
+  if (!NumberMatcher().Matches(symbol, &manager)) return;
+  const auto width_leaf = manager.GetAs<SyntaxTreeLeaf>("width");
+  const auto literal_node = manager.GetAs<SyntaxTreeNode>("literal");
+  if (!width_leaf || !literal_node) return;
+
+  const auto width_text = width_leaf->get().text();
+  size_t width;
+  if (!absl::SimpleAtoi(width_text, &width)) return;
+
+  const auto& base_digit_part = literal_node->children();
+  auto base_leaf = down_cast<const SyntaxTreeLeaf*>(base_digit_part[0].get());
+  auto digits_leaf = down_cast<const SyntaxTreeLeaf*>(base_digit_part[1].get());
+
+  const auto base_text = base_leaf->get().text();
+  const auto digits_text = digits_leaf->get().text();
+
+  const BasedNumber number(base_text, digits_text);
+  int bits_per_digit = 1;
+  switch (number.base) {
+    case 'd':
+      return;  // Don't care about decimal values.
+    case 'b':
+      if (!check_bin_numbers_) return;
+      bits_per_digit = 1;
+      break;
+    case 'o':
+      if (!check_oct_numbers_) return;
+      bits_per_digit = 3;
+      break;
+    case 'h':
+      if (!check_hex_numbers_) return;
+      bits_per_digit = 4;
+      break;
+    default:
+      LOG(FATAL) << "Unexpected base '" << base_text << "'";  // Lexer issue ?
+  }
+
+  // Allow literals with single "0" or "?" as an exception
+  if (width > number.literal.length() * bits_per_digit &&
+      number.literal != "0" && number.literal != "?") {
+    violations_.insert(LintViolation(
+        digits_leaf->get(),
+        FormatReason(width_text, base_text, number.base, digits_text),
+        context));
   }
 }
 
 // Generate string representation of why lint error occurred at leaf
 std::string UndersizedBinaryLiteralRule::FormatReason(
-    absl::string_view width, absl::string_view base,
+    absl::string_view width, absl::string_view base_text, char base,
     absl::string_view literal) {
-  return absl::StrCat("Binary literal ", width, base, literal,
-                      " is shorter than its declared width: ", width, ".");
+  absl::string_view base_describe;
+  switch (base) {
+    case 'b':
+      base_describe = "Binary";
+      break;
+    case 'h':
```

Comment:
do we need case 'x' as well?

Reply:
Is `x` a valid base ? I only know about **b**inary, **d**ecimal, **o**ctal, **h**ex (which is also the set of values that `BasedNumber` can do.